macros/src/codegen.rs

```diff
@@ -526,16 +540,22 @@ pub fn generate_code(sm: &ParsedStateMachine) -> proc_macro2::TokenStream {
             /// It will return `Ok(&NextState)` if the transition was successful, or `Err(Error)`
             /// if there was an error in the transition.
             pub fn process_event(&mut self, #temporary_context mut event: Events) -> Result<&States, Error> {
-                match self.state {
+                match self.state.take().unwrap() {
```

Comment:
Using an option like this may require significant data copying, and I don't see th eneed for having the state wrapped in an option. Is this required?

Reply:
As I explained to your previous comment, it is needed to enable us to get the owned value. In case of match failure I set the state back to `Some`.

Reply:
Actually I just found a bug. I forgot to restore the state if guard fails. I will fix it later

Reply:
This is what the generated code whould be:

```rust
    pub fn process_event(&mut self, mut event: Events) -> Result<&States, Error> {
        match self.state.take().unwrap() {
            States::State2(state_data) => match event {
                _ => {
                    self.state = Some(States::State2(state_data));
                    Err(Error::InvalidEvent)
                }
            },
            States::State1 => match event {
                Events::Event1(event_data) => {
                    match self.context.guard(event_data) {
                        Ok(()) => {
                            let _data = self.context.state1_event1_action(event_data);
                            self.state = Some(States::State2(_data));
                        }
                        Err(e) => {
                            self.state = Some(States::State1);
                            return Err(Error::GuardFailed(e));
                        }
                    }
                    Ok(self.state.as_ref().unwrap())
                }
                _ => {
                    self.state = Some(States::State1);
                    Err(Error::InvalidEvent)
                }
            },
            state => {
                self.state = Some(state);
                Err(Error::InvalidEvent)
            }
        }
    }
```

Reply:
`unwrap` is notorious to make embedded images quite large.

Reply:
I see. Another idea is to have a hidden transitional state (e.g. `States::_Transitional`)  and use `std::mem::replace(&mut self.state, State::_Transitional)` which again will give us an owned value back,

Reply:
I tried the `States::_Transitional` but didn't like it. Mainly because if the user has a match for `States` then they need to handle `States::_Transitional` too.